server/notifications.go

```diff
@@ -5,91 +5,87 @@
 package server
 
 import (
-	"context"
 	"sync"
-
-	"github.com/gofrs/uuid"
-)
-
-const (
-	// how many notifications can be in a notification buffer.
-	notificationBufferSize = 16
+	"sync/atomic"
 )
 
-type change struct {
-	// persistedIndex is a changed persisted index or `nil` if the value has not changed.
-	persistedIndex *int64
+// state is an extendable struct that stores
+// the information subscribers need to be notified about
+type state struct {
+	persistedIndex uint64
 }
 
-// Any returns true if there is a change.
-func (c change) Any() bool {
-	return c.persistedIndex != nil
+// Equals returns true if 2 instances of the state are equal
+func (s *state) Equals(v *state) bool {
+	return (s == nil && v == nil) || (s != nil && v != nil && s.persistedIndex == v.persistedIndex)
 }
 
 type notifications struct {
-	subscribers map[uuid.UUID]chan change
-	mutex       *sync.Mutex
+	mutex *sync.RWMutex
+	cond  *sync.Cond
+
+	// the whole state must be always replaced with `notify`
+	// the struct fields are not thread safe on their own
+	state *state
+
+	// the notifications are stopped when this flag > 0
+	stopped uint32
 }
 
-func (n *notifications) subscribe() (out <-chan change, stop func(), err error) {
-	ch := make(chan change, notificationBufferSize)
-	id, err := uuid.NewV4()
-	if err != nil {
-		return nil, nil, err
-	}
+// getState return the current state or `nil` if the initial state
+// is not set or if the notifications are stopped.
+// This is thread-safe.
+func (n *notifications) getState() *state {
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
+	return n.state
```

Comment:
Is returning a pointer to the state struct thread safe? The underlying value is still mutable.

I doubt there is a performance hit to just returning a copy here.

Reply:
Please see the comment here https://github.com/elastic/elastic-agent-shipper/commit/48d073c16d6b41c77cc2dc527b49c39217224769#diff-67dcdf9d934cf656c54f93e1cd0ab9e971bfab3ccf0188a300d693a6cb190c4cR27-R28

Reply:
You could return `state, bool` where the boolean value indicates if `state` is valid as one possible alternative.